[New Feature] Key Overrides

JonasGessner opened the pull request:
<!--- Provide a general summary of your changes in the title above. -->

<!--- This template is entirely optional and can be removed, but is here to help both you and us. -->
<!--- Anything on lines wrapped in comments like these will not show up in the final text. -->

## Description

I originally pitched this feature in #11301 and on the qmk discord, and now my code is ready to be reviewed, tested and merged. I wrote a detailed documentation, and I will [refer to this document](https://github.com/JonasGessner/qmk_firmware/blob/key-overrides/docs/feature_key_overrides.md) for an explanation of this feature. This document covers the basic idea of the feature, how it is used, why it is different from combos and more.

This seems to be a frequently requested feature (see [here](https://www.reddit.com/r/olkb/comments/5olxit/can_i_make_a_layermodifier_hybrid_key_in_qmk/), [here](https://www.reddit.com/r/olkb/comments/8s115m/qmk_same_modifier_applied_to_different_keys_how/), [here](https://www.reddit.com/r/olkb/comments/covpq3/problem_checking_for_modifier_key_on_custom_key/), #108, #1495), but no solid solution has yet been implemented yet (see #2900, #4795, and my comments about #4795 here: #11301).

Also see #11301 for additional details, including some of the implementation decisions. This PR contains a few core changes, but these are purely additive.

I put a lot of thought and effort into this feature, as will be evident from the documentation, and I think people will really like using it. I have been using and testing it myself for over half a year, and I have been continuously improving it.

While precise benchmarking is hard due to the limited granularity of milliseconds from the timer.h functions, I performed some benchmarks. I measured the `process_key_override` function, which is executed from `process_record_quantum`, and it is clocking in at 0ms-1ms per key press on [my keyboard](https://github.com/JonasGessner/qmk_firmware/tree/jonas/keyboards/xd68/keymaps/jonas), which contains 28 key overrides. These numbers don't provide much insight, but at least they show that key overrides add a minimal overhead. To compare, a call to `layer_on` takes about 10ms on my keyboard.

## Types of Changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply. -->
- [x] Core
- [ ] Bugfix
- [x] New feature
- [ ] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [x] Documentation

## Issues Fixed or Closed by This PR

* #11301

## Checklist

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

dnaq commented:
This is an amazing feature, thanks for the work you have put into it. It will greatly simplify having custom keymaps for different languages.

I have found one issue while trying to port my keymap to it, when the replacement key is a custom keycode, then a bunch of modifiers are added to the keypress. I'll give a minimal example of what I mean:

```
enum custom_keycodes {
  CU_GRV = SAFE_RANGE,
  CU_CIRC,
}

const key_override_t override_6 = ko_make_basic(MOD_MASK_SHIFT, KC_6, CU_CIRC);

const key_override_t **key_overrides = (const key_override_t *[]){ &override_6, NULL};

bool process_record_user(uint16_t keycode, keyrecord_t *record) {
  switch (keycode) {
    case CU_CIRC:
      return false;
    default:
      return true;
  }
```
Here a press of Shift-6 won't yield the keycode CU_CIRC, instead it will yield a weird combination of characters, that seems to be dependent on the custom keycode it has. For example using the above code for me yields the following output (from xbindkeys):
```
"(Scheme function)"
    m:0x11 + c:50
    Shift+Mod2 + Shift_L
"(Scheme function)"
    m:0xd5 + c:105
    Control+Shift+Mod2+Mod4+Mod5 + Control_R
"(Scheme function)"
    m:0xd1 + c:108
    Shift+Mod2+Mod4+Mod5 + ISO_Level3_Shift
"(Scheme function)"
    m:0x51 + c:134
    Shift+Mod2+Mod4 + Super_R
"(Scheme function)"
    m:0x11 + c:21
    Shift+Mod2 + dead_acute
"(Scheme function)"
    m:0x11 + c:50
    Shift+Mod2 + Shift_L
```

The problem I'm trying to solve using your pull request is to have a US keyboard layout on an international keyboard. I have a much more convoluted solution for this in current keymap, but your approach is a lot nicer to work with.

The example has been stripped down to be minimal, but it should be reproducible.

JonasGessner commented:
Thanks for your feedback @dnaq and I'm glad you find this feature useful! I tested your example and reproduced the issue. Turns out I had to add some more checks to make it work properly with custom keycodes. It should work now!

Keep in mind though that `process_record_user` will not be called for your custom keycode when you set it up as replacement key in a key override. With the current design of qmk, `process_record_user` is only called for physical key presses, so it will not be called for the replacement key registered by the key overrides feature. Unfortunately there is no equivalent to `process_record_user` for keycodes registered in software.

Key overrides are designed with this in mind, and they allow you to set up a function that is called every time the override activates and deactivates. If you look at the key overrides docs, there is an example showing how you can register a custom function (it is the advanced example at the bottom), and there is also documentation text on the `custom_action` member of `key_override_t`. Hope this works and any more feedback is welcome!

dnaq commented:
Thanks for the fix. This almost solves my use-case, there's still some code that I need to write to handle dead keys and shifted dead keys (I used a macro for these previously), but it greatly simplifies the keymap at least.

I think a viable solution is to have a custom keycode that checks the modifiers and returns true if the modifiers that the override triggers on are set, causing a wee bit of code duplication, but still much cleaner than my old solution.

treeman commented:
Very cool feature.

Is it possible to get it working with auto-shift? I tried this kind of override (with a Swedish keyboard in the OS):

```
// / -> !
const key_override_t slsh_override = ko_make_basic(MOD_MASK_SHIFT, SE_SLSH, SE_EXLM);
```

JonasGessner commented:
It is possible, but would require modifying the auto shift code. Unfortunately the various features (like auto shift) that are spread throughout the tmk and qmk code are implemented without a unified interface and by different developers, making it really hard to have these different features work together nicely. While there is a clean interface for handling physical key presses (`process_record_*`), there is no equivalent for handling software-based key presses (i.e. calling functions like `register_code`, `add_mods`, etc). Auto shift registers keycodes using these functions, while key overrides only handle physical key presses, since there is no design in place for handling software-based key presses. A fundamental redesign would be necessary here.

So once this PR is accepted, I could look at adding support for auto shift with a separate PR, where I would have to dig into the auto shift code and make it specifically call into the key overrides code.

One comment for your code snippet: The `trigger` keycode has to be without any modifiers, since you specify the modifiers separately (i.e. `MOD_MASK_SHIFT` in your code snippet). `SE_SLSH` is defined as `S(SE_7)`, therefore you should just specify `SE_7`:

```c
// / -> !
const key_override_t slsh_override = ko_make_basic(MOD_MASK_SHIFT, SE_7, SE_EXLM);
```

treeman commented:
Alright, thank you.

As for my code snippet I do have to use `SE_SLSH` as it doesn't seem to work with `SE_7`?

JonasGessner commented:
@treeman ok, I took a look at your keymap and it seems I misunderstood what you are trying to do! You're right to use `SE_SLSH` as trigger keycode.

I thought you had the keycode `SE_7` on your keymap and wanted to map this to `SE_EXLM` when shift is held, but you actually have `SE_SLSH` (`S(SE_7)`) on your keymap, so specifying `SE_SLSH` as trigger keycode is correct in this case.

JonasGessner commented:
PSA: The name of this feature is not set in stone yet. I don't think it's necessarily the best name, so if anyone has ideas for a better name I'd be happy to hear them.

theol0403 commented:
This PR looks amazing! I come from #4795, but this implementation looks much better! I look forward to using this new functionality for all sorts of stuff.

However, I am decently tight for firmware space on my board, but this PR absolutely obliterates my available space and goes 1.2k bytes over (and that's with the removal of #4795). I measured this PR to take about 1.8k bytes on my board. 

Looking at the diff, I see a ton of functionality that I will probably never use. For example: callbacks, negmods, suppressed mods, options, layers, enabled, key override on/off/toggle/keycodes. They still take up a lot of space regardless of whether I use them or not.

My request to you, is to put as much functionality as possible behind macro config options. If I want to use any of the above-listed features, I need to enable a macro in config.h. This will greatly save on space, and keep this new feature exactly as complex as the user wants it.

I think that by default, this PR should only provide `ko_make_basic(modifiers, key, replacement)`, and as you need more advanced features, you can enable them.

I do however see myself using quite a few of these features (I quite like your layer key example), so any other way to reduce firmware space would be greatly appreciated!

Thank you!

JonasGessner commented:
Thanks for your feedback! Code size-wise I don't see a whole lot than can be done without turning off certain features. Your idea of adding extra flags to enable/disable some of the features could certainly be a way to reduce code size, although it would also make the source code itself more bloated and complicated to follow.

Feedback by reviewers would be helpful on this. I could, for example, add a flag that either turns on or turns off the `options`, `negative_mods`, `custom_action`, `context` and `enabled` features. The savings in compiled code size would not be monumental for the feature itself, but it would make the type `key_override_t` smaller. Depending on how many overrides you use, these savings could add up to provide a reasonable amount of size reduction.

I would consider leaving this PR as-is for now, and perhaps adding a flag like `KEY_OVERRIDE_DISABLE_EXTRA` at a later point in a separate PR, which allows turning off the mentioned features. That kind of change would be non-breaking.

Also, you've probably already done this, but you can check this guide for ideas to reduce code size: https://thomasbaart.nl/2018/12/01/reducing-firmware-size-in-qmk/

theol0403 commented:
Sounds good! Given the amount of space the feature uses, without looking at the code, I can imagine much of it is to provide support for the various features mentioned, so could removing the API for various interfaces also remove much of the internal code? Hopefully implementing that won't be too difficult. There could be other optimizations done too, like removing the keyrepeat functionality if the defer delay is 0. 

Thanks again for making this feature however, and I hope the reviewers will have some insight on this problem!

julihoh commented on docs/feature_key_overrides.md:
```diff
@@ -0,0 +1,228 @@
+
+## Difference to Combos
+
+Note that key overrides are very different from [combos](https://docs.qmk.fm/#/feature_combo). Combos require that you press down several keys almost _at the same time_ and can work with any combination of non-modifier keys. Key overrides work like keyboard shortcuts (e.g. `ctrl` + `z`): They take combinations of _multiple_ modifiers and _one_ non-modifier key to then perform some custom action.  ...
```
"... can be used to really fine-tune the $. ..."

$ is missing

99epep commented:
Hi Jonas, thanks for your work !

Is it possible to override a mod-tap key ?
I mean, in these two ways : overriding the key sent by a mod-tap when tapped, and keeping the modifier functionality of a mod-tap on which an override is activated ?

tzarc commented:
Sorry about that, GitHub decided to delete the `develop` branch from upstream when we merged it, despite being told otherwise. Reopened.

JonasGessner commented:
> Hi Jonas, thanks for your work !
> 
> Is it possible to override a mod-tap key ?
> I mean, in these two ways : overriding the key sent by a mod-tap when tapped, and keeping the modifier functionality of a mod-tap on which an override is defined ?

Yes you can use mod tap keys. If you want to use the mod tap key as the trigger key you have to specify the full `MT(.., ..)` keycode as the `trigger` key. If you want to use the modifier state of a mod tap key as one of the required modifiers to activate a key override you don't need to do anything special.

99epep commented:
> Yes you can use mod tap keys. If you want to use the mod tap key as the trigger key you have to specify the full `MT(.., ..)` keycode as the `trigger` key. 

Thanks, works fine !

> If you want to use the modifier state of a mod tap key as one of the required modifiers to activate a key override you don't need to do anything special.

That is not the point, I would like an overrided mod-tap key to still be a mod-tap key. For example I’d like this to be possible : 
`const key_override_t a_ko = ko_make_basic(0, MT(MOD_LSFT, BP_A), MT(MOD_LSFT, FR_A));`

JonasGessner commented:
I see, you want the replacement key to be a mod-tap. That won't work unfortunately! It's a non-trivial case to handle. Getting these different features to work together nicely is very hard and I am not planning on adding support for MT replacement keys.

99epep commented:
Thanks again Jonas, yes keep it as simple as can be ! I just wanted to know as I’m not able to understand this by myself. The trick I asked better suits a layer change.

I’m looking for the easiest way to "override" or "remap" a layout over another one, in my case I want a bépo layout (kind of french Dvorak) when plugged on a french/azerty device. Your key-overrides allow to obtain this without defining a specific layer (unsure it’s clever), unless there’s mod-taps around.

I tried your feature in combinaison with [sevanteri’s early_combo](https://github.com/qmk/qmk_firmware/pull/8591) and everything seems to work as expected.

theol0403 commented:
Hey again, finally made some space on my board to experience the power of this feature :yum:. 

I made a little helper file to make working with overrides even nicer!

The user-facing code looks like this:
```cpp
// simple overrides go here
#define OVERRIDE_DEF                                                \
  OVERRIDE_BASIC(delete_key, MOD_MASK_SHIFT, KC_BSPACE, KC_DELETE)  \
  OVERRIDE_BASIC(tilde_esc, MOD_MASK_SHIFT, KC_ESC, S(KC_GRAVE))    \
  OVERRIDE_LAYERS_NEGMODS_OPTIONS(prev_track_override_example, MOD_MASK_CS, KC_MPLY, KC_MPRV, ~0, MOD_MASK_ALT, ko_option_no_reregister_trigger)

// complicated overrides go here
const key_override_t vol_up_override = ko_make_with_layers_negmods_and_options(MOD_MASK_ALT, KC_MPLY, KC_VOLU, ~0, MOD_MASK_CS, ko_option_no_reregister_trigger);

const key_override_t vol_down_override = ko_make_with_layers_negmods_and_options(MOD_MASK_SA, KC_MPLY, KC_VOLD, ~0, MOD_MASK_CTRL, ko_option_no_reregister_trigger);

// register the complicated overrides
#define OVERRIDE_EXTRAS &vol_up_override, &vol_down_override,

// alternative syntax for the above *could possibly* be:
#define OVERRIDE_EXTRAS                       \
  OVERRIDE_REGISTER_EXTRA(vol_up_override)    \
  OVERRIDE_REGISTER_EXTRA(vol_down_override)

// where the magic happens
#include "override_helper.h"
```

And, `override_helper.h` is
```cpp
// this helper takes an override definition (defined by OVERRIDE_DEF) and converts it into qmk-compatible override syntax

// helper defines which feed into the `OVERRIDE` define 
#define OVERRIDE_BASIC(name, ...) OVERRIDE(name, ko_make_basic(__VA_ARGS__))
#define OVERRIDE_LAYERS(name, ...) OVERRIDE(name, ko_make_with_layers(__VA_ARGS__))
#define OVERRIDE_LAYERS_NEGMODS(name, ...) OVERRIDE(name, ko_make_with_layers_and_negmods(__VA_ARGS__))
#define OVERRIDE_LAYERS_NEGMODS_OPTIONS(name, ...) OVERRIDE(name, ko_make_with_layers_negmods_and_options(__VA_ARGS__))

// for this step we use `OVERRIDE` to create the objects
#define OVERRIDE(name, initializer) static const key_override_t override_##name = initializer;
OVERRIDE_DEF
#undef OVERRIDE

// extra overrides not in `OVERRIDE_DEF` to be included in the array
#ifndef OVERRIDE_EXTRA
#  define OVERRIDE_EXTRA
#endif

// for this step we use `OVERRIDE` to name the overrides to put in the array
#define OVERRIDE(name, initializer) &override_##name,
const key_override_t **key_overrides = (const key_override_t *[]){OVERRIDE_DEF OVERRIDE_EXTRA NULL};
#undef OVERRIDE
```

I think it makes the overrides definition much more concise and easy to maintain!

You are welcome to use or modify this in any way you like, if you wish to include it in the pr or documentation.

Thanks again for this awesome feature!

JonasGessner commented:
I'm no stranger to doing crazy stuff with macros, but it took me a while to wrap my head around this 😅 I think this might be a bit overkill though. Not that I don't want you to use your helper macros if you prefer them — but you don't need to define every override as its own variable and you can get away with the example below. In my opinion it is not as hard to understand and it seems more organized as well. 
```c
// complicated overrides go here (if you want linebreaks, comments, or initializers), like normal use of this PR
const key_override_t vol_up_override = ko_make_with_layers_negmods_and_options(MOD_MASK_ALT, KC_MPLY, KC_VOLU, ~0, MOD_MASK_CS, ko_option_no_reregister_trigger);

const key_override_t vol_down_override = ko_make_with_layers_negmods_and_options(MOD_MASK_SA, KC_MPLY, KC_VOLD, ~0, MOD_MASK_CTRL, ko_option_no_reregister_trigger);

const key_override_t **key_overrides = (const key_override_t *[]){
    // Overrides from your OVERRIDES_DEF macro
    &ko_make_basic(MOD_MASK_SHIFT, KC_BSPACE, KC_DELETE),
    &ko_make_basic(MOD_MASK_SHIFT, MOD_MASK_SHIFT, KC_ESC, S(KC_GRAVE)),
    &ko_make_with_layers_negmods_and_options(MOD_MASK_CS, KC_MPLY, KC_MPRV, ~0, MOD_MASK_ALT, ko_option_no_reregister_trigger),
    // Overrides from your OVERRIDE_EXTRA macro
    &vol_up_override, // These could also be directly defined here
    &vol_down_override,
    NULL};
```

Thanks for testing this feature!

theol0403 commented:
Ah, I didn't think of creating the objects directly inside the array :joy:. In that case, you are totally right, it is overkill!

Wulfsta commented:
Tested this out (here)[https://github.com/Wulfsta/qmk_firmware/tree/wulfsta-overrides] and I'm experiencing some issues.

You'll notice that on [this layer](https://github.com/Wulfsta/qmk_firmware/blob/wulfsta-overrides/keyboards/hhkb/jp/keymaps/wulfsta-key-overrides/keymap.c#L64) I am sending an [ANSI Shifted Key](https://beta.docs.qmk.fm/using-qmk/advanced-keycodes/keycodes_us_ansi_shifted), `KC_LPRN`, that (should be overridden)[https://github.com/Wulfsta/qmk_firmware/blob/wulfsta-overrides/keyboards/hhkb/jp/keymaps/wulfsta-key-overrides/keymap.c#L76] by `KC_LT`. This works most of the time, unless the key is held and shift is toggled on and off; output looks like the following instead:
```
((((((((((((((((((((((999999999999999(((((((((9999999999999((((((((((((9999999999((((((((((9
```
I suspect this could easily be averted with custom keycodes and an override for both states, but this behavior is certainly not intended.

theol0403 commented:
I am experiencing the same issue with `ko_make_basic(MOD_MASK_SHIFT, LT(NUM, KC_BSPC), KC_DELETE)`. Holding the mod-tap and tapping shift causes backspace to be held (even after letting go of everything).

Wulfsta commented:
Custom keycodes did not seem to fix the problem for me, rather than smoothly toggling between the two options when toggling shift it does not continue outputting registered keys. I suspect that this might be due to options that need to be set for all the overrides, but don't know if that's the case.

JonasGessner commented:
> Tested this out [here](https://github.com/Wulfsta/qmk_firmware/tree/wulfsta-overrides) and I'm experiencing some issues.
> 
> You'll notice that on [this layer](https://github.com/Wulfsta/qmk_firmware/blob/wulfsta-overrides/keyboards/hhkb/jp/keymaps/wulfsta-key-overrides/keymap.c#L64) I am sending an [ANSI Shifted Key](https://beta.docs.qmk.fm/using-qmk/advanced-keycodes/keycodes_us_ansi_shifted), `KC_LPRN`, that [should be overridden](https://github.com/Wulfsta/qmk_firmware/blob/wulfsta-overrides/keyboards/hhkb/jp/keymaps/wulfsta-key-overrides/keymap.c#L76) by `KC_LT`. This works most of the time, unless the key is held and shift is toggled on and off; output looks like the following instead:
> 
> ```
> ((((((((((((((((((((((999999999999999(((((((((9999999999999((((((((((((9999999999((((((((((9
> ```
> 
> I suspect this could easily be averted with custom keycodes and an override for both states, but this behavior is certainly not intended.

This is happening because your trigger modifier is shift and the replacement keycode is a shifted keycode. `ko_make_with_layers` sets the `suppressed_mods` to the trigger modifiers. Meaning that since shift if the trigger modifier, the shift key won't be registered, even if it is part of the replacement keycode. That's how the suppressed mods work, they completely prevent the suppressed modifiers from being registered while the override is active.

So to circumvent this you would have to set `suppressed_mods` to 0 for that override. Which you can currently not do with any of the `ko_make` macros. But you could look at how those macros are defined and make a custom one where you can specify `suppressed_mods` or always leave it at 0.

Wulfsta commented:
> > Tested this out [here](https://github.com/Wulfsta/qmk_firmware/tree/wulfsta-overrides) and I'm experiencing some issues.
> > You'll notice that on [this layer](https://github.com/Wulfsta/qmk_firmware/blob/wulfsta-overrides/keyboards/hhkb/jp/keymaps/wulfsta-key-overrides/keymap.c#L64) I am sending an [ANSI Shifted Key](https://beta.docs.qmk.fm/using-qmk/advanced-keycodes/keycodes_us_ansi_shifted), `KC_LPRN`, that [should be overridden](https://github.com/Wulfsta/qmk_firmware/blob/wulfsta-overrides/keyboards/hhkb/jp/keymaps/wulfsta-key-overrides/keymap.c#L76) by `KC_LT`. This works most of the time, unless the key is held and shift is toggled on and off; output looks like the following instead:
> > ```
> > ((((((((((((((((((((((999999999999999(((((((((9999999999999((((((((((((9999999999((((((((((9
> > ```
> > 
> > 
> > I suspect this could easily be averted with custom keycodes and an override for both states, but this behavior is certainly not intended.
> 
> This is happening because your trigger modifier is shift and the replacement keycode is a shifted keycode. `ko_make_with_layers` sets the `suppressed_mods` to the trigger modifiers. Meaning that since shift if the trigger modifier, the shift key won't be registered, even if it is part of the replacement keycode. That's how the suppressed mods work, they completely prevent the suppressed modifiers from being registered while the override is active.
> 
> So to circumvent this you would have to set `suppressed_mods` to 0 for that override. Which you can currently not do with any of the `ko_make` macros. But you could look at how those macros are defined and make a custom one where you can specify `suppressed_mods` or always leave it at 0.

I mostly fixed this [here](https://github.com/Wulfsta/qmk_firmware/blob/0d8420a3d5aadba1e8d13ac75c0a68e4ff2db95b/keyboards/hhkb/jp/keymaps/wulfsta-key-overrides/keymap.c), how do I modify the behavior of toggling shift on and off? Doing so appears to unregister the codes with my current firmware.

JonasGessner commented:
> I am experiencing the same issue with `ko_make_basic(MOD_MASK_SHIFT, LT(NUM, KC_BSPC), KC_DELETE)`. Holding the mod-tap and tapping shift causes backspace to be held (even after letting go of everything).

Interoperability with things like `LT` is not perfect and you may run into issues. I've already complained a few times how qmk features are implemented all over the place and that it is really hard to make different features work together nicely, sadly! Basically, the code that handles `LT` only acts when the `LT` key is _physically_ pressed or lifted. With key overrides, however, you want to unregister keys that are still physically held. qmk has this strict distinction between physical key presses (`process_record_xxx`) and software key presses (`(un)register_code`). All key overrides can use is `(un)register_code16`, but sending an `LT` code to this function does nothing and hence all sorts of problems ensue! `LT` is only handled in `process_record_xxx`, but I can't call this function without it being a massive hack that would never get approved. There should be a single interface for registering and unregistering keycodes, so that physical and simulated key events can work together better.

theol0403 commented:
Humm, I'm not quite sure I follow, that's too bad, but it seems to work alright most of the time! The problem is not when I hold shift and press the LT, the problem is when I hold the LT to move to the `NUM` layer, and then from inside that layer I tap `KC_SHIFT`. That's when it starts being all weird. Maybe you already understood me, but I'm just clarifying in case this is actually a bug that can be fixed.

skullydazed commented:
> There should be a single interface for registering and unregistering keycodes, so that physical and simulated key events can work together better.

FWIW we're well aware of this, but that decision pre-dates QMK and the work involved in changing it is considerable. It's high on my wishlist of things to address.

JonasGessner commented:
> 
> I mostly fixed this [here](https://github.com/Wulfsta/qmk_firmware/blob/0d8420a3d5aadba1e8d13ac75c0a68e4ff2db95b/keyboards/hhkb/jp/keymaps/wulfsta-key-overrides/keymap.c), how do I modify the behavior of toggling shift on and off? Doing so appears to unregister the codes with my current firmware.

I looked at the code you linked – you are setting the `negative_mod_mask` to shift now, this won't help. You need to use the `suppressed_mods` field in `key_override_t`. Please check the [docs](https://github.com/JonasGessner/qmk_firmware/blob/key-overrides/docs/feature_key_overrides.md) they have details on all of this :)

What you want is this:

```c
const key_override_t l1_kc_lprn_override = {.trigger_mods      = MOD_MASK_SHIFT,      //
                                            .layers            = NUMPAD_LAYER,        //
                                            .suppressed_mods   = 0,                   //  <---!
                                            .options           = ko_options_default,  //
                                            .negative_mod_mask = 0,                   //
                                            .custom_action     = NULL,                //
                                            .context           = NULL,                //
                                            .trigger           = KC_LPRN,             //
                                            .replacement       = KC_LT,               //
                                            .enabled           = NULL};
```

If you have to do this for several overrides I recommend defining a custom macro à la `ko_make_xxx` macros.

Wulfsta commented:
@JonasGessner Heh, that's actually what I was going to do at first, but I thought I'd get the same behavior with the negative mask. Guess I'll fix it up, thanks!

JonasGessner commented:
> > There should be a single interface for registering and unregistering keycodes, so that physical and simulated key events can work together better.
> 
> FWIW we're well aware of this, but that decision pre-dates QMK and the work involved in changing it is considerable. It's high on my wishlist of things to address.

Yeah I totally see how this separation happened, with the tmk core and then qmk additions on top and the highly distributed nature of the project, etc. I wasn't trying to direct blame to anyone!

Getting this done seems like an incredible effort, so kudos for considering.

JonasGessner commented:
> Tested this out [here](https://github.com/Wulfsta/qmk_firmware/tree/wulfsta-overrides) and I'm experiencing some issues.
> 
> You'll notice that on [this layer](https://github.com/Wulfsta/qmk_firmware/blob/825fadacc54a5d6c8f2847c3c85678d59f26e588/keyboards/hhkb/jp/keymaps/wulfsta-key-overrides/keymap.c#L64) I am sending an [ANSI Shifted Key](https://beta.docs.qmk.fm/using-qmk/advanced-keycodes/keycodes_us_ansi_shifted), `KC_LPRN`, that [should be overridden](https://github.com/Wulfsta/qmk_firmware/blob/825fadacc54a5d6c8f2847c3c85678d59f26e588/keyboards/hhkb/jp/keymaps/wulfsta-key-overrides/keymap.c#L76) by `KC_LT`. This works most of the time, unless the key is held and shift is toggled on and off; output looks like the following instead:
> 
> ```
> ((((((((((((((((((((((999999999999999(((((((((9999999999999((((((((((((9999999999((((((((((9
> ```
> 
> I suspect this could easily be averted with custom keycodes and an override for both states, but this behavior is certainly not intended.

I went back and took a look at this and realized that the issue you were having here has nothing to do with key overrides. 

First of all, while what I said about having to set suppressed_mods is true, I forgot that I built this feature in a way where you do not need to explicitly change suppress_mods to 0 when the replacement keycode contains the same modifier as the trigger modifiers. It is handled automatically. So you will be able to just get away with what you were doing originally (https://github.com/Wulfsta/qmk_firmware/blob/825fadacc54a5d6c8f2847c3c85678d59f26e588/keyboards/hhkb/jp/keymaps/wulfsta-key-overrides/keymap.c). Just go back to that code, it will work the same without being unnecessarily complicated.

Second, you set up the `l3_o_rprn_override` to only activate on layer 3. But you also have `KC_LPRN` on layer 0, so the key override does not activate there unless you set the override to also activate on layer 0. Check the docs for how to control on which layers overrides activate.

And third, the issue you were having was happening on layer 0, where the key override never activated. This issue happens without any key overrides, and with the key overrides feature completely turned off even. If you have `KC_LPRN` on  a key, hold down the key and then press and release shift a couple of times you will start getting output like`(((((99999((99((((99`. I believe weak modifiers automatically get unregistered when another key is pressed down, which would explain exactly this misbehavior. You should definitely file a bug report for this and this is definitely not correct. You can just take this paragraph as a template for the bug report.

So to hopefully wrap up with your bugs, revert your code back to [this](https://github.com/Wulfsta/qmk_firmware/blob/825fadacc54a5d6c8f2847c3c85678d59f26e588/keyboards/hhkb/jp/keymaps/wulfsta-key-overrides/keymap.c) and make sure you correctly set the layers for the overrides. If you do not want to restrict an override to specific layers then just use `ko_make_basic` which makes the override work on any layer.

Wulfsta commented:
> > Tested this out [here](https://github.com/Wulfsta/qmk_firmware/tree/wulfsta-overrides) and I'm experiencing some issues.
> > You'll notice that on [this layer](https://github.com/Wulfsta/qmk_firmware/blob/825fadacc54a5d6c8f2847c3c85678d59f26e588/keyboards/hhkb/jp/keymaps/wulfsta-key-overrides/keymap.c#L64) I am sending an [ANSI Shifted Key](https://beta.docs.qmk.fm/using-qmk/advanced-keycodes/keycodes_us_ansi_shifted), `KC_LPRN`, that [should be overridden](https://github.com/Wulfsta/qmk_firmware/blob/825fadacc54a5d6c8f2847c3c85678d59f26e588/keyboards/hhkb/jp/keymaps/wulfsta-key-overrides/keymap.c#L76) by `KC_LT`. This works most of the time, unless the key is held and shift is toggled on and off; output looks like the following instead:
> > ```
> > ((((((((((((((((((((((999999999999999(((((((((9999999999999((((((((((((9999999999((((((((((9
> > ```
> > 
> > 
> > I suspect this could easily be averted with custom keycodes and an override for both states, but this behavior is certainly not intended.
> 
> I went back and took a look at this and realized that the issue you were having here has nothing to do with key overrides.
> 
> First of all, while what I said about having to set suppressed_mods is true, I forgot that I built this feature in a way where you do not need to explicitly change suppress_mods to 0 when the replacement keycode contains the same modifier as the trigger modifiers. It is handled automatically. So you will be able to just get away with what you were doing originally (https://github.com/Wulfsta/qmk_firmware/blob/825fadacc54a5d6c8f2847c3c85678d59f26e588/keyboards/hhkb/jp/keymaps/wulfsta-key-overrides/keymap.c). Just go back to that code, it will work the same without being unnecessarily complicated.
> 
> Second, you set up the `l3_o_rprn_override` to only activate on layer 3. But you also have `KC_LPRN` on layer 0, so the key override does not activate there unless you set the override to also activate on layer 0. Check the docs for how to control on which layers overrides activate.
> 
> And third, the issue you were having was happening on layer 0, where the key override never activated. This issue happens without any key overrides, and with the key overrides feature completely turned off even. If you have `KC_LPRN` on a key, hold down the key and then press and release shift a couple of times you will start getting output like`(((((99999((99((((99`. I believe weak modifiers automatically get unregistered when another key is pressed down, which would explain exactly this misbehavior. You should definitely file a bug report for this and this is definitely not correct. You can just take this paragraph as a template for the bug report.
> 
> So to hopefully wrap up with your bugs, revert your code back to [this](https://github.com/Wulfsta/qmk_firmware/blob/825fadacc54a5d6c8f2847c3c85678d59f26e588/keyboards/hhkb/jp/keymaps/wulfsta-key-overrides/keymap.c) and make sure you correctly set the layers for the overrides. If you do not want to restrict an override to specific layers then just use `ko_make_basic` which makes the override work on any layer.

The behavior I was seeing was all happening on layer 3, or at least with the layer 3 key held while toggling right shift on and off. I very specifically only want this to happen on layer 3, which is why I wrote the code that way. You’ll observe in a latter commit where I used custom keycodes that I use the basic function to override a custom keycode for escape to 'KC_ESC' on all layers.

JonasGessner commented:
The actaul cause of the `(((((99999((99((((99` like output is still not key overrides. Turn off key overrides completely and you'll see it still happen.

I looked at your code again. The reason the override is still not activating is because the shift key is on layer 0 in your keymap, and the override will only activate for key events originating from layer 3. Layer 3 is transparent where shift is, so shift actually comes from layer 0. So you'd have to move things around a bit, for example simply redefine shift on layer 3 instead of having a transparent keycode.

Wulfsta commented:
> The actaul cause of the `(((((99999((99((((99` like output is still not key overrides. Turn off key overrides completely and you'll still see it happen.
> 
> I looked at your code again. The reason the override is still not activating is because the shift key is on layer 0 in your keymap, and the override will only activate for key events originating from layer 3. Layer 3 is transparent where shift is, so shift actually comes from layer 0. So you'd have to move things around a bit, for example simply redefine shift on layer 3 instead of having a transparent keycode.

I will give this a try, it was not clear to me that a transparent key over a modifier would not cause the layer override to activate.

rmurrish commented:
it looks like the check that fails is just the travis job timing out. 

> The job exceeded the maximum time limit for jobs, and has been terminated.

is there a way around that to pass the check and get this merged?

Erovia reviewed (changes requested):
Very nice feature and I think a lot of users would like to use it.
Just a few things:

* needs a rebase to fix the conflict
* documentation should link to the **docs.qmk.fm** domain, not the beta. one
* as someone who just checked out the documentation for the feature (like a new user would/will), I think the paragraphs are not in the best order, my suggestion would be:

    * Key Overrides
    * Setup
    * Creating Key Overrides
    * Keycodes
    * Simple Example
    * Intermediate Examples
    * Advanced Examples
    * Inner Workings
    * Difference to Combos
    * Reference for key_override_t
    * Reference for ko_option_t

JonasGessner commented:
Thanks for the review, I went ahead and fixed the conflict and the links.

I agree that the docs weren't ordered ideally. Especially the "Difference to Combos" section is not that important and I moved it to the very end, while moving all examples together and moving "Keycodes" up. The "Inner Workings" section heavily relies on the type reference for `key_override_t` so I'd like to keep the "Inner Workings" section after the "Reference for key_override_t" section.

JonasGessner commented:
I also didn't mean to remove the review request from qmk/collaborators, all I did was request a new review from @Erovia 🤔 I can't request a review from qmk/collaborators, could you please do this @Erovia ?

tzarc reviewed (changes requested):
Haven't tested personally, but looks okay in principle. Just a small change required.

fauxpark commented on quantum/quantum_keycodes.h:
```diff
@@ -409,6 +409,11 @@ enum quantum_keycodes {
     KC_LAPO,  // 5CF6
     KC_RAPC,  // 5CF7
 
+    // Key Overrides
+    KEY_OVERRIDE_TOGGLE,
+    KEY_OVERRIDE_ON,
+    KEY_OVERRIDE_OFF,
+    
```
Worth noting everything including and following the Sequencer keycodes will not be able to be supported by VIA, because the Sequencer keycodes do not have fixed values. See #11156

JonasGessner replied:
Rebased and moved the keycodes

JonasGessner commented:
Can this still make it into the current develop branch?

drashna commented on common_features.mk:
```diff
@@ -318,6 +318,11 @@ ifeq ($(strip $(PRINTING_ENABLE)), yes)
     SRC += $(TMK_DIR)/protocol/serial_uart.c
 endif
 
+ifeq ($(strip $(KEY_OVERRIDE_ENABLE)), yes)
```
Please add an entry to `show_options.mk`, as well.

drashna commented on common_features.mk:
```diff
@@ -318,6 +318,11 @@ ifeq ($(strip $(PRINTING_ENABLE)), yes)
 
+ifeq ($(strip $(KEY_OVERRIDE_ENABLE)), yes)
+    OPT_DEFS += -DKEY_OVERRIDE_ENABLE
+    SRC += $(QUANTUM_DIR)/key_override.c
```
Based on what this is doing, I feel that moving the files to the `process_keycode` folder would be very appropriate here. 

```suggestion
    SRC += $(QUANTUM_DIR)/process_keycode/key_override.c
```

drashna commented on quantum/quantum.c:
```diff
@@ -58,12 +58,17 @@ float bell_song[][2] = SONG(TERMINAL_SOUND);
 
-static void do_code16(uint16_t code, void (*f)(uint8_t)) {
+#ifdef KEY_OVERRIDE_ENABLE
+extern bool process_key_override(const uint16_t keycode, const keyrecord_t *const record);
+extern void matrix_scan_key_override(void);
```
These should be in a header file for the feature (`key_override.h`), and not prototyped here.

JonasGessner replied:
The reason I didn't expose these functions in `key_override.h` is because they are meant to be used only internally in quantum and should not be exposed in keymaps and userspace (to avoid people calling these functions because they pop up in autocompletion). I'll move them to a new header `key_override_private.h`.

drashna replied:
A good number of other similar features are in the same boat, but still expose the stuff. 

But this should be okay, I think.

drashna commented:
Also, could you add the function here:
https://github.com/qmk/qmk_firmware/blob/master/docs/understanding_qmk.md#process-record

baffalop commented:
This looks like a great feature, thank you! Actually I can't believe that with all the features QMK has this is only just a PR — it feels like a natural extension.

I just wanted to suggest a couple of things regarding the negmods option. This might be coloured by the use case that I intend for the feature, but before I had read as far as the negmods initialisers and fully absorbed the implications, I would've assumed that the modifiers would be exclusive. For example, currently I want to move my quote into semicolon position, but still get colon when I type it shifted. But if I'm entering some keyboard shortcut involving quote, eg `cmd + shift + '`, I would still want it to register as such rather than be translated to `cmd + shift + ;`.

So two suggestions:
- I think it'd be very helpful to make clear early on in the documentation that this is _not_ how `ko_make_basic` works by default, rather than having it tacitly implied by the presence of the negmods option.
- Maybe provide another initialiser, such as `ko_make_exclusive`, that sets `negative_mods` to be `~trigger_mods`?

JonasGessner commented:
Thanks for the feedback! I made the docs on `ko_make_basic` clearer regarding what you mentioned, but I'll keep the code as it is. If more demand for a macro like `ko_make_exclusive` comes up this can always easily be added later.

rmurrish commented:
FYI, that CI error is on includes for a specific keyboard, that was reportedly fixed in #13108.

JonasGessner commented:
Pinging because cut off for next breaking changes merge is in one month. Would be nice to see this make it for that merge.

drashna commented:
Could you run `qmk cformat` on the files in question?

JonasGessner commented:
The remaining linter errors are on code that this PR doesn't touch.

zvecr commented on quantum/process_keycode/process_key_override.c:
```diff
@@ -0,0 +1,518 @@
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include QMK_KEYBOARD_H
```
```suggestion
```

JonasGessner replied:
Replaced with `#import "quantum.h"`

drashna commented:
Thanks!

Ekleog commented:
Hello people! I'm new to QMK (currently basing my config off 52ce2088e8b84fd888f91824d2ff30bd98d1e232) and trying this out (to make “é” capitalize to “É” with my keyboard set as azerty, which means send caps-lock / é / caps-lock so it doesn't give the default “2”).

The way I tried to do it was by using a custom key that outputs “É” with the above-mentioned command sequence in process_record_user, and then use the following code to set my key override:
```c
const key_override_t seacu_key_override = ko_make_basic(MOD_MASK_SHIFT, FR_EACU, EK_SEACU);
const key_override_t **key_overrides = (const key_override_t *[]){
    &seacu_key_override,
    NULL
};
```

I first checked by having a key have the `EK_SEACU` effect that my process_record_user did what I want. It looks like it works.

But with the key override above, when I try to press shift+eacu, here is what happens:
- I press shift, X gets “shift pressed”
- I press and release eacu, X gets “shift release” “shift pressed” (I assume as a result of the key override)
- I release shift, X gets “shift released”

So it looks like my custom keycode's `tap_code` calls aren't getting output.

Is it expected that this feature doesn't work with custom keycodes? If no, what am I doing wrong? (Bonus points if there's an easier way to do what I'm trying to do)

Anyway, thank you for QMK, I'm still very new to it (flashed for the first time yesterday) but it looks awesome to me already, and this PR looks like exactly what I was looking for!

Ekleog commented:
Oh. I was pointed to the answer, and well, looks like it was hidden in the github-hidden comments, sorry for the noise! The answer is `ko_make_basic` can't do this and it must go through the full syntax, as explained in https://github.com/qmk/qmk_firmware/pull/11422#issuecomment-754720883

Chikowitz commented:
> Thanks for the fix. This almost solves my use-case, there's still some code that I need to write to handle dead keys and shifted dead keys (I used a macro for these previously), but it greatly simplifies the keymap at least.
> 
> I think a viable solution is to have a custom keycode that checks the modifiers and returns true if the modifiers that the override triggers on are set, causing a wee bit of code duplication, but still much cleaner than my old solution.

@dnaq hello, can you maybe share an example of how did you solve your problem? I think I'm trying to do the same.
